### ISSUE TITLE
Display land ownership on map.bitquest.co

### ISSUE THREAD

**Issue by jpchip** [bitquest/bitquest#97]
It would be really cool to be able to see land ownership on the map or a version of the map. 

It appears the map is generated via [Minecraft Overviewer](https://overviewer.org/), which has the ability to [add custom Points of Interest in a config file](http://docs.overviewer.org/en/latest/signs/#manual-pois). Could this config file, which is in Python. query the redis database to get the chunk owners? Or alternatively maybe create a script that queries the database and generates the config file itself with the chunk ownership in it? 

Looking for feedback or ideas or just general thoughts on feasibility.

**Comment by Xeyler** [comment]
I think this is a great idea, but I admittedly don't have much experience with the Minecraft Overviewer. It does look like you did your homework well, though.

I am concerned with how these points of interest are displayed, though. If I'm correct, they're displayed as small icons on the map, but if we had the map make one icon for every claim, it would be not only informative, but also very hard to read under the many claim icons. 

Perhaps we could make the icons not display until a player toggles them on, or zooms in really far? Again, I don't know much about the overviewer, but I think this would be a good idea.

**Comment by jpchip** [comment]
Thanks for the feedback. I totally agree we need to allow users to toggle the layer, and to limit the zoom. 

I did a little more research, and I see two possible approaches (2 is my favorite):

1. I found an existing bukkit plugin, [PoiManager](https://dev.bukkit.org/bukkit-plugins/poimanager/), which does a lot of what would be trying to do, but isn't quite right. It gave me an idea though how we could go about generating the list of points from within the bitquest plugin and make it accessible to Minecraft Overviewer though. I bet if we got access to the source code we could tweak it for our purposes. It also looks like we can set the `checked` flag on the manual POI definition to false so that the points aren't on by default, but I think we'd have to do some tinkering to limit the zoom. 

2. Doing this as points was just my first thought, since it seemed easiest based on the Minecraft Overviewer docs. But what I really want to do was create an actual overlay, if it is possible. But last night I found an example: [Wynncraft Map](https://map.wynncraft.com)! Their Guild Territory overlay is pretty close to what I was thinking, and looking at the code it's not hard to implement on the Overviewer side.

I could create the javascript script to be added to the Minecraft Overviewer index.html that create the ownership overlay, with a on/off toggle and a zoom limit. 

What I'd need someone else with some java chops to do (if this is even possible) is create a simple service endpoint that returns the list of owner chunks with their coordinates as JSON, based on the current bounds. Something like this:

`GET http://play.bitquest.co/api/chunks?startX"=-736&startZ"=-4428&endX"=-802"&endZ"=-4553`

```
[{
"owner" : "jpchip",
"name" : "my land",
"location": {"startX":-740,"startZ":-4429,"endX":-756,"endZ":-4445}
},
{
"owner" : "xeyler",
"name" : "some other land",
"location": {"startX":-757,"startZ":-4445,"endX":-773,"endZ":-4461}
},
...
]
```

So, is 2 even an option? I'm open to other thoughts and suggestions too.

**Comment by Xeyler** [comment]
This is very interesting. You've certainly worked on it quite a bit, and I think it can work.

If I'm correct, you want a service that responds to HTTP requests with JSON containing all the chunk ownerships within certain coordinates. 

Honestly, this really excites me. If we could get this service running, we could allow anyone to access it, enabling client side mods that can display information about chunks around them. Also, if we could make the chunk information overlay able to be turned on or off, I would fully support adding it to the map.

I may be able to supply the Java chops, but I'll probably need to recruit @explodi for help along the way.

**Comment by jpchip** [comment]
You are correct, I want a service that responds to HTTP requests with JSON containing all the chunk ownerships within certain coordinates. That is what I was really asking if it was even possible. 

I am sure we can get the overlays to appear, I've worked with the Google Maps API quite a bit and looking at how the [Wynncraft code](https://map.wynncraft.com/js/territoryLocations.js) works it's pretty straightforward, **IF** we have that service. 

That is a good point about the skewing though. In the example, the boundaries are being projected into the isometric coordinate system when they drawn, so they are close but it's not going to be perfect. Basically I wouldn't use this map to settle border disputes in court, but it would be good enough for our purposes. 

You can actually see how a chunk will look by going to http://map.bitquest.co/, opening chrome dev tools, and in the console tab pasting in the following code:

```
function getLatLng(x, z) {
  var coords = overviewer.util.fromWorldToLatLng(x, 64, z, overviewer.mapView.options.currentTileSet);
  
  return {
    lat: coords.lat(),
    lng: coords.lng()
  };
}

var startX = 0;
var endX = 16;
var startZ = 0;
var endZ = -16;

var paths = [];
paths.push(getLatLng(startX,startZ));
paths.push(getLatLng(endX,startZ));
paths.push(getLatLng(endX,endZ));
paths.push(getLatLng(startX,endZ));

var polygon = new google.maps.Polygon({
  paths: paths,
  strokeColor: '#fff',
  strokeOpacity: 0.8,
  strokeWeight: 2,
  fillColor: '#fff',
  fillOpacity: 0.35
});

polygon.setMap(overviewer.map);
```

Change those start and end variables as desired to see chunks in different locations.

**Comment by Xeyler** [comment]
I've taken a look at the map with some overlays, and while it looks OK, I think there might be a way to make the overlays more accurate.

I posted an issue on the Minecraft overviewer GitHub, explaining this issue, and the response was interesting. Take a look at the discussion [here](https://github.com/overviewer/Minecraft-Overviewer/issues/1332).

Apparently, he's using the slime chunks visualizer to get that effect, but perhaps there's a way to add overlays that look like that and include the owner and land name data. He says that all overlays look like that, but apparently not... Perhaps we need to change some settings to get that effect...

**Comment by CounterPillow** [comment]
You're using the region feature currently, not overlays.

Overlays get rendered out at map render time into image tiles, and are a separate layer that can be overlaid onto the image tiles.

**Comment by jpchip** [comment]
@CounterPillow, is it possible to define a custom rendermode to generate an overlay from a list of chunk boundaries (eg. `{"startX":0,"startZ":0,"endX":16,"endZ":-16}`) at map render time then? I can't figure out how one would do that looking at the [docs](http://docs.overviewer.org/en/latest/config/#the-rendermode-primitives).

**Comment by CounterPillow** [comment]
@jpchip no, you have to write your own overlay rendermode in C.

**Comment by Xeyler** [comment]
@CounterPillow, thanks so much for taking a look at this discussion. 

I must admit that I'm not personally very familiar with the overviewer, but would it be possible to, at render, generate many different overlays with many different colors(All in C, obviously)?

Optimally, we would want to be able to display other data about the overlayed parts in a way like POIs, but if that's not possible, we could simply use overlays and POIs in tandem.

**Comment by CounterPillow** [comment]
>but would it be possible to, at render, generate many different overlays with many different colors(All in C, obviously)?

Yes, and one overlay can also have multiple colours (such as the mineral overlay)

>Optimally, we would want to be able to display other data about the overlayed parts in a way like POIs, but if that's not possible, we could simply use overlays and POIs in tandem.

Pretty sure overlays work on a per-block basis, so you can't easily do things like add text to them. Using POIs for that might be the better approach.

**Comment by jpchip** [comment]
So to recap, I guess we are down to: 

1. just using markers to show the land ownership, 
2. javascript based overlays that don't line up well and need a JSON service, or
3. forking Minecraft Overviewer and writing a custom renderer in C that takes bounding coordinates so that the overlays are generated at map run time directly on the blocks, then adding markers on top with the names.

**Comment by Xeyler** [comment]
Yup, that's where we are as of now. 

I think the best thing we can do is configure a custom renderer to make accurate overlays. Then in JavaScript, we can define POIs at the South-West corners of each claim

If I read the documentation correctly, POIs can take Y coordinates as well, so if we include the full coordinates of the highest block in the south-west corner of each chunk, we can avoid the height issue that JavaScript regions causes.

After this, **I really think we should make a BitQuest API**(perhaps on `api.bitquest.co`?). It could include claim data as well as Bitcoin wallet information, player information, server information, the works. It would be extremely helpful for moderators, and even more helpful for developers.

**Comment by jpchip** [comment]
Based on our discussions I forked Minecraft Overviewer and started work on a custom rendermode primitive to color overlays on given bounding coordinates:

https://github.com/jpchip/Minecraft-Overviewer/tree/overlay-bounds

The idea is in the config file you can define a list of (((minx, minz, maxx, maxz), ...), (r, g, b, a)) tuples for coloring boundaries. That way we list out all the chunks a user owns and give them a color. For example:

```
BoundsOverlay(bounds=[
    (((0, -16, 16, 0)), (255, 160, 122, 255)),
    (((32, 32, 48, 48), (128, 128, 144, 144)), (255, 255, 0, 255)),
])]
```

The problem is I don't actually know python OR C, so I need a little help. :p 

I created an [overlay-bounds.c](https://github.com/jpchip/Minecraft-Overviewer/blob/overlay-bounds/overviewer_core/src/primitives/overlay-bounds.c) class, blatantly ripped off from [overlay-structure.c](https://github.com/jpchip/Minecraft-Overviewer/blob/overlay-bounds/overviewer_core/src/primitives/overlay-structure.c). The code compiles fine ([in windows](https://github.com/overviewer/Minecraft-Overviewer/issues/1333) no less!), but when I try to run it on my local world it dies. If I take don't use the rendermode it works fine, so I definitely have a problem either in my overlay-bounds.c or in the bounds I'm passing in. 

Could someone (@theraccoonbear?) code review for me, and tell me if I'm doing something stupid? Once it's working I will do a pull request to try to get it added to the original project.

**Comment by Xeyler** [comment]
Impressive work, @jpchip.

I must admit, though, that I don't know much of Python or C either. I've suggested a BitQuest API to @explodi, and he has said that we may be able to set something up with enough time. Of course, doing this with maximum security will be a challenging and perhaps lengthy process, but hopefully it can be effective.

In the meantime, **we may be able to use REDIS** from the class written in C within the Minecraft Overviewer to get the claim data straight from REDIS into our overviewer(which means we can move on without an API just yet). [REDIS](http://redis.io/) is a database we use for storing all sorts of variables(including claim data), and it [supports both Java and C](http://redis.io/clients), but I'm not sure if two different processes can access the same REDIS database at the same time.

I'm not sure, but I believe @explodi knows enough C to pull this off.

**Comment by explodi** [comment]
I will start working on a API prototype this weekend. I can also upload the docker container used to display bitquest map


Cristian

**Comment by explodi** [comment]
Here is the source code (docker image) we are using for the BitQuest map.

https://github.com/bitquest/overviewer-docker

Feel free to send a pull request. It was meant to be a "general purpose" overviewer image for Minecrafters but perhaps we can fork it and create the bitquest-overviewer repository (that talks to the future BitQuest API)

If you guys think this is the way to go. Let me know

**Comment by Xeyler** [comment]
Well, @explodi, this whole discussion is probably TL;DR by now, but we mentioned that it would require us to create a custom rendermode in C to get the high-quality renders for less confusing overlays. It looks like the overviewer docker you set up just has scripts, so we may need to completely fork the original minecraft overviewer source into our own repository, but @CounterPillow would probably be able to tell us more on that than I could.

Also, I would like to emphasize the API. To be honest, I think it would be nice to be able to see some analytics about BitQuest. Naturally, we would leave personal things out such as bitcoin address keys, IPs, etc, but there are things I'd like to include that aren't stored in REDIS. For example, I'd like to make available a list of all the players that have ever played BitQuest, all of their statistics, their levels, the locations of plots they own, and a whole bunch of statistics on the world such as generated chunks, live entities, total days, basically anything about everything.

This does raise the question about how much should be considered private material. Well, if this was on api.bitquest.co, and it was returned in the form of JSON, I don't believe it would get a lot of popularity beyond our moderators and developers, and as long as we leave out bitcoin keys and IPs, everything should be ok.

TL;DR I(and possibly @jpchip?) have a lot of things I want added to the API, so we should make a list and talk about if these things are possible, @explodi

**Comment by CounterPillow** [comment]
>so we may need to completely fork the original minecraft overviewer source into our own repository, but @CounterPillow would probably be able to tell us more on that than I could.

Just create a patch and apply it to upstream git HEAD. Saves you from keeping the fork up to date all the time.

**Comment by jpchip** [comment]
I believe I have the bounding box overlay in Minecraft Overviewer working. If anyone would like to code review before I merge into my local master, the pull request is here:

https://github.com/jpchip/Minecraft-Overviewer/pull/1

Once merged, I will do a pull request to try to get it added to the original project, if they are interested. Otherwise we could follow CounterPillow's advice, or just fork it into the bitquest group.

**Comment by Xeyler** [comment]
Hey, it looks like you've got some pretty nice looking overlays working, @jpchip. After a quick look, it seems like this should do pretty well.

Hopefully, I'll have more time later to check the code a little more. For now, do you think it would be possible to easily implement a JSON response from a BitQuest API to specify overlays? Hopefully, we'll be able to get a simple API up and running soon.

Also, we should probably set the opacity of the overlays to about 50% for visibility.

**Comment by jpchip** [comment]
> do you think it would be possible to easily implement a JSON response from a BitQuest API to specify overlays?

That I'm not so sure about, not sure even what language to do it in! I understand the end goal would be a script to call the BitQuest API and parse the results to generate a config file as [documented here](http://docs.overviewer.org/en/latest/faq/#i-want-to-put-manual-poi-definitions-or-other-parts-of-my-config-into-a-seperate-file), which would define the list of bounding boxes AND corner markers. I'm just not sure about implementation...

**Comment by theraccoonbear** [comment]
An API built into the BQ code is appealing, but really this could just be a one-off script that talks to Redis and generates the needed land ownership data and is run on a cron, or whatever scheduler is currently used, when the current Overviewer process runs every day.  @explodi may have thoughts on the best way to incorporate this into the map generation workflow.

**Comment by Xeyler** [comment]
I suspect integration with REDIS would be best because it already holds the information for claims. I think we should mostly be worried about security, though. If we do access REDIS for claim details, it will need to be in a completely secure way because it's also where we store sensitive bitcoin information.

That said, I'm not very familiar with the minecraft overviewer, C, or JavaScript, so I doubt I'll be able to be very helpful. @explodi is probably going to have to implement some things server-side to properly and securely communicate between the overviewer and REDIS, and it seems that @jpchip and @theraccoonbear will be valuable in organizing the code necessary to integrate the overviewer with REDIS. Should we organize assignments?

**Comment by Burntuno** [comment]
The idea of a map overlay of claimed land is great.  One thing to worry about is, should players know who owns what?  Being ambiguous like Bitcoin, gives the Bitquest World more mystery.  An overlay of just the plot names would keep this more enigmatic and give players a chance to reveal as needed.

**Comment by juanj** [comment]
I opened a PR to https://github.com/bitquest/overviewer-docker/pull/1 to add your code as a patch @jpchip.
I also made [this script](https://github.com/juanj/bitquest-overviewer-land-config) that connects to redis and generate a config file with the BoundsOverlay.

**Comment by CubeFromSpace** [comment]
Is this entirely essential? I'm not criticizing or anything, but I think that may be too much because of all the points showing and blocking out part of the map

**Comment by jpchip** [comment]
It's not essential at all, I just thought it would be a neat feature. 

The way I ended up implementing it is not with points at all, but with a [colored overlay](https://github.com/jpchip/Minecraft-Overviewer/pull/1). You can also toggle the overlay layer on or off.

**Comment by CubeFromSpace** [comment]
Ok, that makes more sense, especially with clans and public/private ownership too.

**Comment by Shiny-andKittens** [comment]
Maybe different colors for Private, Public, and Clan
